### PR TITLE
Implement sharing of an absorption correction in scaling

### PR DESCRIPTION
**Pull request by jbeilstenedmands**
In order to compare absorption on multi-orientation datasets, a common crystal frame of reference needs to be used (a prerequisite for refining a shared absorption surface).

This is achieved by applying the inverse of `(S * R * F)` to the lab-frame `s0` and `s1` vectors. Before, the z-coordinate was used to determine a phi value, which doesn't give consistency across sweeps for multi-axis goniometers. The change on the overall scaling result _should be_ negligible as we're just using different reference frames.

As it is difficult to interpret the absorption correction, several more plots have been added:
e.g. for beta-lactamase, scaling with absorption_level=medium, anomalous=True

'Absorption correction surface', as before
newplot (45)
This plot is not particularly useful on its own, as there is no indication of where it is being sampled by the scattering vectors.

'Undiffracted absorption correction' i.e. absorption for a beam passing straight through, which gives an idea of the 'shape' of the crystal:
newplot (46)

'Scattering vectors in the crystal frame', useful to see where scattering is actually occurring in this reference frame.
newplot (49)

The distribution of the actual absorption corrections applied.
newplot (48)

For high absorption cases, the absorption correction surface can refine to extreme values, but this can be misleading as the actual relative absorption corrections applied are much more modest.

I think these new plots help to see whats going on and I would welcome feedback on further improvements.

### CODE REVIEW

**Comment by jbeilstenedmands** [comment]
Update: I have also implemented sharing of the absorption correction between sweeps for the physical scaling model, with the option `share.absorption=True`. I decided to include this in this PR as it makes the original changes more useful and makes the whole change set easier to test on interesting datasets. I have tested on some user data from I19, turning on the absorption correction sharing results in the overall metrics being ever so slightly worse, likely due to the loss of a high number of free parameters from the scaling models, but I think this is more than acceptable for models which should now be 'more correct'.

**Review by benjaminhwilliams — APPROVED** [review]
Generally looks good.  I have a list of changes I would recommend for more efficient calculation of the spherical harmonics, but I'll make a separate PR for that.

File: algorithms/scaling/plots.py [hunk 1]
```diff
@@ -451,6 +460,191 @@ def plot_absorption_surface(physical_model):
...
+                        r = sqrt2 * ((-1) ** m) * Ylm.imag
+                    elif m == 0:
+                        assert Ylm.imag == 0.0
+                        r = Ylm.real
+                    else:
+                        r = sqrt2 * ((-1) ** m) * Ylm.real
+                    Intensity[ip, it] += 0.5 * params[counter] * r
+                    p2 = (p + np.pi) % (2.0 * np.pi)
+                    t2 = np.pi - t
+                    Ylm = nsssphe.spherical_harmonic(l, abs(m), t2, p2)
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply algorithms/scaling/plots.py:498]: On reflection, feel free to ignore this suggestion, I'll make a separate PR.

**Comment by benjaminhwilliams** [comment]
Just looking again at the plots, I realise that the _θ_ and _φ_ labels reflect the mathematician's convention for spherical coordinates, whereby _φ_ ∈ (0, _π_] is the zenith and _θ_ ∈ (0, 2_π_] is the azimuth.  I always instinctively use the opposite, physicist's convention, and it looks like the `nsssphe.spherical_harmonic` method does too (as does the SciPy equivalent), so we end up with opposite conventions in the code and in the plots.  Perhaps, to avoid confusion between the two, or between these angles and goniometer axes, would it be worth just labelling them as 'zenith' (or 'polar angle') and 'azimuth'?

**Comment by jbeilstenedmands** [comment]
I've adopted most of the suggestions from @benjaminhwilliams, thanks. If we want to use the scipy spherical harmonics, I'd rather change to this in a separate PR for both instances in the scaling code  in the model and the plotting, to ensure consistency as there are many peculiarities to watch out for in the `nsssphe` code. 
If the tests now pass, happy for this to be merged.

**Comment by ndevenish** [comment]
```
Traceback (most recent call last):
  File "/Users/runner/work/1/build/../modules/dials/command_line/scale.py", line 258, in <module>
    run()
  File "/Users/runner/work/1/conda_base/python.app/Contents/lib/python3.6/contextlib.py", line 52, in inner
    return func(*args, **kwds)
  File "/Users/runner/work/1/build/../modules/dials/command_line/scale.py", line 231, in run
    scaled_experiments, joint_table = run_scaling(params, experiments, reflections)
  File "/Users/runner/work/1/build/../modules/dials/command_line/scale.py", line 192, in run_scaling
    algorithm.run()
  File "/Users/runner/work/1/modules/dials/algorithms/scaling/algorithm.py", line 231, in run
    logger.info("%s%s%s", "\\n", "=" * 80, "\\n")
  File "/Users/runner/work/1/modules/dials/algorithms/scaling/observers.py", line 189, in __exit__
    _make_scaling_html(self.script)
  File "/Users/runner/work/1/modules/dials/algorithms/scaling/observers.py", line 200, in _make_scaling_html
    make_scaling_model_plots(scaling_script.experiments, scaling_script.reflections)
  File "/Users/runner/work/1/modules/dials/algorithms/scaling/observers.py", line 251, in make_scaling_model_plots
    scaling_model_plots = plot_scaling_models(data[key], reflection_tables[key])
  File "/Users/runner/work/1/modules/dials/algorithms/scaling/model/model.py", line 1401, in plot_scaling_models
    return model.plot_model_components(reflection_table=reflection_table)
  File "/Users/runner/work/1/modules/dials/algorithms/scaling/model/model.py", line 931, in plot_model_components
    d.update(plot_absorption_plots(self, reflection_table))
  File "/Users/runner/work/1/modules/dials/algorithms/scaling/plots.py", line 572, in plot_absorption_plots
    Intensity[x, y] = 1
IndexError: Please report this error to dials-support@lists.sourceforge.net: index 180 is out of bounds for axis 1 with size 180
```
Do you understand the source of these?

**Comment by jbeilstenedmands** [comment]
@ndevenish looking into it right now

**Pull request merged**